projects/FW_104/src/fw_104_can.c

```diff
@@ -0,0 +1,102 @@
+#include "can.h"
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+
+#include "fw_104_can.h"
+
+#define CAN_SEND_TIME_MS 1000
+
+StatusCode prv_can_callback(const CanMessage *msg, void *context, CanAckStatus *ack_reply) {
+  // Taken from smoke_can
+  LOG_DEBUG("Received a message!\n");
+  printf("Data:\n\t");
+  for (uint8_t i = 0; i < msg->dlc; i++) {
+    uint8_t byte = (msg->data >> (i * 8)) & 0xFF;
+    printf("0x%x ", byte);
+  }
+  printf("\n");
+  return STATUS_CODE_OK;
+}
+
+void prv_can_transmit_A(SoftTimerId timer_id, void *context) {
+  uint8_t messageData = 5;
+  CanMessage message = {
+    .source_id = 0x1, .msg_id = 0x1, .data_u8 = { messageData }, .type = CAN_MSG_TYPE_DATA, .dlc = 8
+  };
+  can_transmit(&message, NULL);
```

Comment:
The 0xA and 0xB in the homework refers to the message ID; the source ID should just match the device ID you initialized with. DLC is the data-length-code, it's the length of the body in bytes (so if you only want to transmit 1 uint8, set it to 1). And to request an ACK, pass a CanAckRequest * (defined in can_ack.h) instead of NULL to can_transmit. For the CanAckRequest, the callback is called when enough ACK messages are received, and the `expected_bitset` is a bitset of the device IDs you want acknowledgement messages from.

20-second rundown on ACKs: when a message is received that is considered "critical" (message id < 13), the CAN library will send back an "ACK" message which has the same message ID but has the ACK bit in the arbitration ID toggled on. The `ack_request` parameter to `can_transmit` tells the CAN library to watch for ACKs from certain device IDs (the expected bitset) and call a callback when those device IDs ACK it. This lets us make sure that certain boards receive certain critical messages.

Reply:
I'm a bit confused with the function of the device id, source id, and message id. For the sake of this module, should the instance sending message id 0xA have the same device id/source id as the instance sending message id 0xB?